File: src/workmux_multirepo/workspace.py
"""Persistent changes: plan, provision, and inspect repository worktrees."""
import fcntl
import hashlib
import json
import re
from contextlib import contextmanager
from pathlib import Path
from .runtime import Error, atomic_json, git, run


def change_id(branch):
    slug = re.sub(r"[^A-Za-z0-9_-]", "-", branch)[:48]
    return f"{slug}-{hashlib.sha256(branch.encode()).hexdigest()[:10]}"


def directory(config, branch):
    return config.state / change_id(branch)


@contextmanager
def locked(config):
    config.state.mkdir(parents=True, exist_ok=True)
    with (config.state / ".lock").open("w") as file:
        try:
            fcntl.flock(file, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError:
            raise Error("Another wmm mutation is running in this workspace.", "Retry after that command finishes.") from None
        yield


def load(config, branch):
    path = directory(config, branch) / "manifest.json"
    if not path.is_file():
        raise Error(f"Change not found: {branch}", "Run wmm status to list changes, or wmm start <branch> <repos...>.")
    try:
        data = json.loads(path.read_text())
    except (ValueError, OSError):
        raise Error("Cannot read the workspace manifest.", f"Restore {path} from a backup; worktrees have not been changed.") from None
    if data.get("version") != 1 or data.get("branch") != branch:
        raise Error("Unsupported or mismatched workspace manifest.")
    return data


def save(config, manifest):
    atomic_json(directory(config, manifest["branch"]) / "manifest.json", manifest)


def plan(config, branch, aliases):
    if not aliases or len(aliases) != len(set(aliases)):
        raise Error("Choose at least one repository without duplicate aliases.", "Run wmm start <branch> <repo> [repo...].", 2)
    unknown = set(aliases) - config.repos.keys()
    if unknown:
        raise Error(f"Unknown repositories: {', '.join(sorted(unknown))}", f"Configured aliases: {', '.join(config.repos)}", 2)
    if len({config.repos[a] for a in aliases}) != len(aliases):
        raise Error("Two selected aliases point to the same repository.", "Select each repository once.", 2)
    if branch.startswith("-") or git(config.root, "check-ref-format", "--branch", branch, check=False).returncode:
        raise Error(f"Invalid branch: {branch}", "Use a valid Git branch such as feat/my-change.", 2)
    records = []
    commons = set()
    for alias in aliases:
        repo = config.repos[alias]
        result = git(repo, "rev-parse", "--show-toplevel", check=False)
        if result.returncode or Path(result.stdout.strip()).resolve() != repo:
            raise Error(f"Not a repository root: {repo}", "Correct the repository path in wmm.toml.")
        common = git(repo, "rev-parse", "--path-format=absolute", "--git-common-dir").stdout.strip()
        if common in commons:
            raise Error("Selected paths share the same Git repository.", "Select each Git repository once.", 2)
        commons.add(common)
        if git(repo, "show-ref", "--verify", "--quiet", f"refs/heads/{branch}", check=False).returncode == 0:
            raise Error(f"Branch already exists in {alias}: {branch}", "Choose a new branch; wmm never adopts unrelated existing branches.")
        # Dry-run deliberately does not contact the network or change refs.
        base = git(repo, "symbolic-ref", "--short", "refs/remotes/origin/HEAD", check=False).stdout.strip()
        records.append(dict(alias=alias, source=str(repo), base=base or "origin/<remote-default>",
                            base_commit=None, path=None, handle=f"wmm-{change_id(branch)}-{alias}", ready=False))
    return records


def fetch_bases(records):
    for record in records:
        repo = record["source"]
        remote = git(repo, "ls-remote", "--symref", "origin", "HEAD").stdout
        match = re.search(r"^ref: refs/heads/(.+)\s+HEAD$", remote, re.MULTILINE)
        if not match:
            raise Error(f"Remote default branch unavailable for {record['alias']}.", "Configure origin with a default branch, then retry wmm start.")
        default = match.group(1)
        git(repo, "fetch", "origin", f"+refs/heads/{default}:refs/remotes/origin/{default}")
        git(repo, "symbolic-ref", "refs/remotes/origin/HEAD", f"refs/remotes/origin/{default}")
        record["base"] = f"origin/{default}"
        record["base_commit"] = git(repo, "rev-parse", f"refs/remotes/origin/{default}^{{commit}}").stdout.strip()


def worktree_for(record, branch):
    text = git(record["source"], "worktree", "list", "--porcelain", "-z").stdout
    for entry in text.split("\0\0"):
        fields = dict(part.split(" ", 1) for part in entry.split("\0") if " " in part)
        if fields.get("branch") == f"refs/heads/{branch}":
            return fields.get("worktree")
    return None


def verify(record, branch):
    path = record.get("path")
    actual = worktree_for(record, branch)
    if not path or not actual or Path(actual).resolve() != Path(path).resolve():
        raise Error(f"Worktree missing or branch changed for {record['alias']}.", "Restore the recorded worktree and branch before retrying; see wmm status <branch>.")


def provision(config, manifest):
    workspace = directory(config, manifest["branch"])
    for record in manifest["repos"]:
        if record["ready"]:
            verify(record, manifest["branch"])
        else:
            actual = worktree_for(record, manifest["branch"])
            if actual:
                # A hook may have failed after Git created the worktree. Do not silently
                # adopt it or call it ready: provisioning completion must be explicit.
                raise Error(f"Incomplete provisioning in {record['alias']}: {actual}",
                            "Inspect the failed setup. Remove that incomplete worktree after preserving any work, then retry wmm start.")
            result = run(["workmux", "add", manifest["branch"], "--headless", "--json",
                          "--name", record["handle"], "--base", record["base_commit"]], cwd=record["source"])
            try:
                receipt = json.loads(result.stdout)
                path = Path(receipt["worktree_path"]).resolve()
                if receipt["schema_version"] != 1 or receipt["branch"] != manifest["branch"]:
                    raise ValueError()
            except (ValueError, KeyError, TypeError):
                raise Error("Unsupported provisioning receipt.", "Use a workmux version with headless JSON schema version 1.") from None
            record["path"] = str(path)
            verify(record, manifest["branch"])
            record["ready"] = True
            save(config, manifest)
        link = workspace / record["alias"]
        if link.is_symlink():
            if link.resolve() != Path(record["path"]):
                raise Error(f"Workspace link points elsewhere: {link}")
        elif link.exists():
            raise Error(f"Workspace path already exists: {link}")
        else:
            link.symlink_to(record["path"], target_is_directory=True)
    manifest["phase"] = "ready"
    save(config, manifest)


def brief(config, manifest):
    workspace = directory(config, manifest["branch"])
    path = workspace / "BRIEF.md"
    if not path.exists():
        rows = "\n".join(f"- {r['alias']}: {r['path']} (base {r['base']} at {r['base_commit']})" for r in manifest["repos"])
        path.write_text(f"# {manifest['branch']}\n\n## Objective\n\n{manifest['objective']}\n\n## Repositories\n\n{rows}\n\n## Contracts and implementation handoff\n\nRecord cross-repository contracts, verification, and remaining questions here before review.\n")
    return path


def status_rows(manifest, fields):
    rows = []
    for record in manifest["repos"]:
        path = record.get("path")
        row = dict(repo=record["alias"], state="incomplete", commits=0, base=record["base"],
                   path=path or "", base_commit=record["base_commit"] or "")
        if record["ready"] and path:
            try:
                verify(record, manifest["branch"])
                dirty = git(path, "status", "--porcelain").stdout
                row["state"] = "modified" if dirty else "clean"
                row["commits"] = int(git(path, "rev-list", "--count", f"{record['base_commit']}..HEAD").stdout)
            except Error:
                row["state"] = "missing-or-switched"
        rows.append({key: row[key] for key in fields})
    return rows

File: internal/cli/workspace.go
package cli

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
)

type repoState struct {
	Alias      string `json:"alias"`
	Source     string `json:"source"`
	Base       string `json:"base"`
	BaseCommit string `json:"base_commit"`
	Path       string `json:"path"`
	Handle     string `json:"handle"`
	Ready      bool   `json:"ready"`
}
type manifest struct {
	Version   int         `json:"version"`
	Branch    string      `json:"branch"`
	Phase     string      `json:"phase"`
	Objective string      `json:"objective"`
	Session   string      `json:"session"`
	Repos     []repoState `json:"repos"`
}

func digest(s string, n int) string {
	sum := sha256.Sum256([]byte(s))
	return hex.EncodeToString(sum[:])[:n]
}
func changeID(branch string) string {
	slug := regexp.MustCompile(`[^A-Za-z0-9_-]`).ReplaceAllString(branch, "-")
	if len(slug) > 48 {
		slug = slug[:48]
	}
	return slug + "-" + digest(branch, 10)
}
func workspace(c *config, branch string) string { return filepath.Join(c.State, changeID(branch)) }
func save(c *config, m *manifest) error {
	return writeJSON(filepath.Join(workspace(c, m.Branch), "manifest.json"), m)
}
func load(c *config, branch string) (*manifest, error) {
	data, err := os.ReadFile(filepath.Join(workspace(c, branch), "manifest.json"))
	if err != nil {
		return nil, fail("Change not found: "+branch, "Run wmm status to list changes, or wmm start <branch> <repos...>.")
	}
	var m manifest
	if err = json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	if m.Version != 1 || m.Branch != branch || len(m.Repos) == 0 {
		return nil, fail("Unsupported or mismatched workspace manifest.", "Restore the manifest from a backup; existing worktrees were preserved.")
	}
	return &m, nil
}
func lock(c *config) (func(), error) {
	if err := os.MkdirAll(c.State, 0755); err != nil {
		return nil, err
	}
	f, err := os.OpenFile(filepath.Join(c.State, ".lock"), os.O_CREATE|os.O_RDWR, 0600)
	if err != nil {
		return nil, err
	}
	if err = syscall.Flock(int(f.Fd()), syscall.LOCK_EX|syscall.LOCK_NB); err != nil {
		f.Close()
		return nil, fail("Another wmm mutation is running in this workspace.", "Retry after that command finishes.")
	}
	return func() { syscall.Flock(int(f.Fd()), syscall.LOCK_UN); f.Close() }, nil
}
func (a *app) plan(c *config, branch string, aliases []string) ([]repoState, error) {
	if len(aliases) == 0 {
		return nil, usage("Choose at least one repository.", help("start"))
	}
	seen := map[string]bool{}
	for _, alias := range aliases {
		if c.Repos[alias] == "" {
			return nil, usage("Unknown repository: "+alias, "Configured aliases: "+strings.Join(keys(c.Repos), ", "))
		}
		if seen[alias] {
			return nil, usage("Duplicate repository: "+alias, help("start"))
		}
		seen[alias] = true
	}
	if strings.HasPrefix(branch, "-") {
		return nil, usage("Invalid branch: "+branch, help("start"))
	}
	if _, err := a.tryGit(c.Root, "check-ref-format", "refs/heads/"+branch); err != nil {
		return nil, usage("Invalid branch: "+branch, help("start"))
	}
	var records []repoState
	commons := map[string]bool{}
	for _, alias := range aliases {
		source := c.Repos[alias]
		root, err := a.tryGit(source, "rev-parse", "--show-toplevel")
		if err != nil {
			return nil, fail("Not a repository root: "+source, "Correct the path in wmm.toml.")
		}
		real, err := canonical(root)
		if err != nil || real != source {
			return nil, fail("Not a repository root: "+source, "Correct the path in wmm.toml.")
		}
		common, err := a.git(source, "rev-parse", "--path-format=absolute", "--git-common-dir")
		if err != nil {
			return nil, err
		}
		if commons[common] {
			return nil, usage("Selected paths share the same Git repository.", "Select each repository once.")
		}
		commons[common] = true
		if _, err = a.tryGit(source, "show-ref", "--verify", "--quiet", "refs/heads/"+branch); err == nil {
			return nil, fail("Branch already exists in "+alias+": "+branch, "Choose a new branch; wmm never adopts unrelated existing branches.")
		}
		base, _ := a.tryGit(source, "symbolic-ref", "--short", "refs/remotes/origin/HEAD")
		if base == "" {
			base = "origin/<remote-default>"
		}
		records = append(records, repoState{Alias: alias, Source: source, Base: base, Handle: "wmm-" + changeID(branch) + "-" + alias})
	}
	return records, nil
}
func (a *app) fetchBases(records []repoState) error {
	pattern := regexp.MustCompile(`(?m)^ref: refs/heads/(.+)\s+HEAD$`)
	for i := range records {
		r := &records[i]
		text, err := a.git(r.Source, "ls-remote", "--symref", "origin", "HEAD")
		if err != nil {
			return err
		}
		match := pattern.FindStringSubmatch(text)
		if len(match) != 2 {
			return fail("Remote default branch unavailable for "+r.Alias, "Configure origin with a default branch, then retry wmm start.")
		}
		branch := match[1]
		if _, err = a.git(r.Source, "fetch", "origin", "+refs/heads/"+branch+":refs/remotes/origin/"+branch); err != nil {
			return err
		}
		if _, err = a.git(r.Source, "symbolic-ref", "refs/remotes/origin/HEAD", "refs/remotes/origin/"+branch); err != nil {
			return err
		}
		r.Base = "origin/" + branch
		r.BaseCommit, err = a.git(r.Source, "rev-parse", "refs/remotes/origin/"+branch+"^{commit}")
		if err != nil {
			return err
		}
	}
	return nil
}
func (a *app) worktreeFor(r repoState, branch string) (string, error) {
	text, err := a.git(r.Source, "worktree", "list", "--porcelain", "-z")
	if err != nil {
		return "", err
	}
	for _, entry := range strings.Split(text, "\x00\x00") {
		var path, ref string
		for _, line := range strings.Split(entry, "\x00") {
			key, value, ok := strings.Cut(line, " ")
			if ok {
				if key == "worktree" {
					path = value
				}
				if key == "branch" {
					ref = value
				}
			}
		}
		if ref == "refs/heads/"+branch {
			return path, nil
		}
	}
	return "", nil
}
func (a *app) verify(r repoState, branch string) error {
	actual, err := a.worktreeFor(r, branch)
	if err != nil {
		return err
	}
	if actual != "" && r.Path != "" {
		left, e1 := canonical(actual)
		right, e2 := canonical(r.Path)
		if e1 == nil && e2 == nil && left == right {
			return nil
		}
	}
	return fail("Worktree missing or branch changed for "+r.Alias, "Restore the recorded worktree and branch before retrying; see wmm status <branch>.")
}
func (a *app) provision(c *config, m *manifest) error {
	for i := range m.Repos {
		r := &m.Repos[i]
		if r.Ready {
			if err := a.verify(*r, m.Branch); err != nil {
				return err
			}
		} else {
			actual, err := a.worktreeFor(*r, m.Branch)
			if err != nil {
				return err
			}
			if actual != "" {
				return fail("Incomplete provisioning in "+r.Alias+": "+actual, "Inspect the failed setup. Preserve any work before removing the incomplete worktree, then retry wmm start.")
			}
			text, err := a.command(r.Source, "workmux", "add", m.Branch, "--headless", "--json", "--name", r.Handle, "--base", r.BaseCommit)
			if err != nil {
				return err
			}
			var receipt struct {
				Version int    `json:"schema_version"`
				Branch  string `json:"branch"`
				Path    string `json:"worktree_path"`
			}
			if err = json.Unmarshal([]byte(text), &receipt); err != nil || receipt.Version != 1 || receipt.Branch != m.Branch || receipt.Path == "" {
				return fail("Unsupported provisioning receipt.", "Use a workmux release with headless JSON schema version 1.")
			}
			r.Path, err = canonical(receipt.Path)
			if err != nil {
				return err
			}
			if err = a.verify(*r, m.Branch); err != nil {
				return err
			}
			r.Ready = true
			if err = save(c, m); err != nil {
				return err
			}
		}
		link := filepath.Join(workspace(c, m.Branch), r.Alias)
		info, err := os.Lstat(link)
		if err == nil {
			if info.Mode()&os.ModeSymlink == 0 {
				return fail("Workspace path already exists: "+link, "Move the conflicting file before retrying.")
			}
			real, e := canonical(link)
			if e != nil || real != r.Path {
				return fail("Workspace link points elsewhere: "+link, "Restore the workspace link before retrying.")
			}
		} else if os.IsNotExist(err) {
			if err = os.Symlink(r.Path, link); err != nil {
				return err
			}
		} else {
			return err
		}
	}
	if m.Phase == "provisioning" {
		m.Phase = "ready"
	}
	return save(c, m)
}
